### PR TITLE
gstreamer: enable webrtc plugin by default

### PR DESCRIPTION
**Pull request by razvanphp**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

This PR enables gstreamer webrtc plugin by default.

### Motivation
Since 2021 webrtc is recommended by W3C for web real-time communication, assigning multiple IETF standards for the same purpose. This plugin is a part of the GStreamer project since early 2018, when webrtcbin was merged and become part of the GStreamer 1.14 release. Since then, the team at Centricular has been working on improving the plugin and adding new features. This plugin is now considered stable and is used by many projects, including popular open-source WebRTC servers.

### Issues

I've read this [comment](https://github.com/Homebrew/homebrew-core/pull/161361#pullrequestreview-1864019540) from another attempt, but I *think* I solved it easier, since `libnice` compiles without `gstreamer` dependancy.

### Tests

I've compiled both gstreamer and libnice locally and tested with brew test & a local app I have developed in python, all is working well and gstreamer finds the necessary plugins (webrtcbin, libnice) on it's own from the brew plugins folder, using the latest version of `libnice` (which is good).

Screenshot 2024-09-07 at 23 22 35

### References

- https://webrtchacks.com/webrtc-plumbing-with-gstreamer/
- https://github.com/Homebrew/homebrew-core/issues/25649
- https://github.com/Homebrew/homebrew-core/issues/151373#issuecomment-1766469175
- https://github.com/Homebrew/homebrew-core/pull/25680
- https://github.com/Homebrew/homebrew-core/pull/161361
- https://github.com/Homebrew/homebrew-core/pull/171902
- https://github.com/Homebrew/homebrew-core/pull/125996#discussion_r1166405225

CC @aconchillo @carlocab @ystreet

### CODE REVIEW

File: Formula/lib/libnice.rb [hunk 1]
```diff
@@ -20,13 +20,14 @@ class Libnice < Formula
     sha256               x86_64_linux:   "392d7eb8bea2d013695a4d57d7a5849d702d018f9d1e2009accd01edbc979784"
   end
 
+  keg_only "it conflicts with gstreamer, which depends on and also includes libnice"
```
Comment (carlocab) [inline Formula/lib/libnice.rb:23]: I don't think making this keg-only is a good solution.

File: Formula/g/gstreamer.rb [hunk 2]
```diff
@@ -77,6 +77,7 @@ class Gstreamer < Formula
   depends_on "json-glib"
   depends_on "lame"
   depends_on "libass"
+  depends_on "libnice"
```
Comment (carlocab) [inline Formula/g/gstreamer.rb:80]: This has no linkage with `libnice` so it doesn't actually seem to be used. The build is likely just building its own `libnice`, which is incompatible with https://docs.brew.sh/Acceptable-Formulae#stuff-that-requires-vendored-versions-of-homebrew-formulae.

**Review by carlocab — CHANGES_REQUESTED** [review]
Thanks, but this approach doesn't work either.

I think you'll need to make the approach at https://github.com/Homebrew/homebrew-core/pull/171902 work instead.

**Comment by razvanphp** [comment]
> Thanks, but this approach doesn't work either.
> 
> I think you'll need to make the approach at #171902 work instead.

Thank you for the quick review!

Unfortunately I think my limited knowledge in meson builds probably makes me the wrong guy for the task, but looking forward for the gstreamer to hop-in hopefully. @nirbheek maybe?

Cheers,
R

**Comment by nirbheek** [comment]
> Unfortunately my limited knowledge in meson builds probably makes me the wrong guy for the task, but looking forward for the gstreamer to hop-in hopefully. @nirbheek maybe?

Unfortunately I know next-to-nothing about Brew, but I am happy to answer any questions you might have, and help with patches to meson build files in libnice or gstreamer.

**Comment by nirbheek** [comment]
That's what https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/6871 is for. Here's what you need to do:

1. Add `-Dgstreamer=disabled` to `libnice.rb`
2. Apple [that patch](https://gitlab.freedesktop.org/gstreamer/gstreamer/-/merge_requests/6871) to `gstreamer.rb`
3. Add  `-Dlibnice-gstreamer-only=true` to the options in `gstreamer.rb`. In this mode, gstreamer will download and build libnice only for the gstreamer plugin, and will use the library libnice from the system.

That's it, this breaks the circular dependency.

**Comment by razvanphp** [comment]
There is a conflict on `meson_options.txt` in that MR.

I did the required changes.

To be honest, I hope for a cleaner solution on the gstreamer build side. I think this should be the default, check first if there is a libnice installed already via `pkg-config` instead of duplicating the build job.

File: Formula/lib/libnice.rb (hunk 1 above)
Comment (razvanphp) [reply Formula/lib/libnice.rb]: reverted.

**Comment by nirbheek** [comment]
> I think this should be the default, check first if there is a libnice installed already via pkg-config instead of duplicating the build job.

The gstnice plugin is not a part of the gstreamer source code. It's part of the libnice source code. GStreamer already does what you describe when pulling in libnice for the webrtc library.

**Comment by carlocab** [comment]
You need to fix the `brew style` failures. Try running `brew style gstreamer libnice` with this branch checked out at `"$(brew --repo homebrew/core)"`. Some of them might be fixed automatically with `brew style --fix gstreamer libnice`.

Also, please split your commits so that each commit only modifies one formula at a time.

> 3. Add  `-Dlibnice-gstreamer-only=true` to the options in `gstreamer.rb`. In this mode, gstreamer will download and build libnice only for the gstreamer plugin, and will use the library libnice from the system.

This will break when we start blocking formulae from downloading things at build-time, so it would be good to avoid downloading `libnice` during a `meson` invocation. This is a big part of why I took the approach I did in #171902.

**Comment by nirbheek** [comment]
> so it would be good to avoid downloading `libnice` during a `meson` invocation

If formulae can specify a fetch step manually, you can just run `meson subprojects download libnice`. During setup, you should pass `meson setup -Dwrap-mode=nodownload` so that meson doesn't download anything automatically.

**Comment by carlocab** [comment]
This is done in https://github.com/Homebrew/homebrew-core/pull/171902.

You'll need to
```
brew update
brew upgrade gstreamer
brew install libnice-gstreamer
```
to pick up the changes. (You might need to wait 20 minutes or so for the formulae.brew.sh to pick up the update.)

**Pull request closed without merging**